### PR TITLE
[Enhancement] Use statistics to optimize skew join

### PR DESCRIPTION
**Pull request by Youngwb**
## Why I'm doing:
In this #34854, we support skew join optimize using shew hint, but it needs user to find skew value, It's not convenient. And we already support collect histogram stats for olap/hive table, we can use most common value(mcv) to find skew value for column.
## What I'm doing:
if join on predicate column has histogram stats, check if it has skewed value, and use the mcv to optimize skew join

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java [hunk 1]
```diff
@@ -1844,6 +1847,15 @@ public SessionVariable setHiveTempStagingDir(String hiveTempStagingDir) {
     @VarAttr(name = SKEW_JOIN_RAND_RANGE, flag = VariableMgr.INVISIBLE)
     private int skewJoinRandRange = 1000;
 
+    @VarAttr(name = USE_STATS_TO_OPTIMIZE_SKEW_JOIN)
+    private boolean useStatsToOptimizeSkewJoin = true;
+
+    @VarAttr(name = SKEW_JOIN_OPTIMIZE_USE_MCV_COUNT, flag = VariableMgr.INVISIBLE)
+    private int skewJoinOptimizeUseMCVCount = 5;
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:1854]: what is the meaning of MCV?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java [hunk 2]
```diff
@@ -642,6 +644,12 @@ private OptExpression pushDownAggregation(OptExpression tree, TaskContext rootTa
         return tree;
     }
 
+    private void skewJoinOptimize(OptExpression tree, TaskContext rootTaskContext) {
+        SkewJoinOptimizeRule rule = new SkewJoinOptimizeRule();
+        Utils.calculateStatistics(tree, rootTaskContext.getOptimizerContext());
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java:649]: what if collect statistics failed, should we do SkewJoinOptimizeRule?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/SkewJoinOptimizeRule.java [hunk 3]
```diff
@@ -85,14 +89,91 @@ public class SkewJoinOptimizeRule extends TransformationRule {
...
 
     @Override
     public boolean check(OptExpression input, OptimizerContext context) {
-        return ((LogicalJoinOperator) input.getOp()).getJoinHint().equals(JoinOperator.HINT_SKEW);
+        // respect the join hint
+        if (((LogicalJoinOperator) input.getOp()).getJoinHint().equals(JoinOperator.HINT_SKEW)) {
+            return true;
+        }
+
+        if (!context.getSessionVariable().isUseStatsToOptimizeSkewJoin()) {
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/SkewJoinOptimizeRule.java:105]: it is better to support both manual and statistics based modes.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/SeriallyTaskScheduler.java [hunk 4]
```diff
@@ -36,7 +36,7 @@ public static TaskScheduler create() {
 
     @Override
     public void executeTasks(TaskContext context) {
-        long timeout = context.getOptimizerContext().getSessionVariable().getOptimizerExecuteTimeout();
+        long timeout = 1000000000000L;
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/SeriallyTaskScheduler.java:39]: reset it back.

**Review by ABingHuang — CHANGES_REQUESTED** [review]

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java (hunk 1 above)
Comment (Youngwb) [reply fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:1854]: Most common value

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java (hunk 2 above)
Comment (Youngwb) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java:649]: Yes, if user use hint, we still need to use SkewJoinOptimizeRule

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/SkewJoinOptimizeRule.java (hunk 3 above)
Comment (Youngwb) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/SkewJoinOptimizeRule.java:105]: Hint is not manual?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/SeriallyTaskScheduler.java (hunk 4 above)
Comment (Youngwb) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/SeriallyTaskScheduler.java]: Done

**Review by ABingHuang — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java [hunk 5]
```diff
@@ -1846,6 +1849,16 @@ public SessionVariable setHiveTempStagingDir(String hiveTempStagingDir) {
     @VarAttr(name = SKEW_JOIN_RAND_RANGE, flag = VariableMgr.INVISIBLE)
     private int skewJoinRandRange = 1000;
 
+    @VarAttr(name = USE_STATS_TO_OPTIMIZE_SKEW_JOIN)
+    private boolean useStatsToOptimizeSkewJoin = true;
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:1853]: enableAutoOptimizeSkewJoin？

Comment (Youngwb) [reply fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java]: Done

**Review by kangkaisen — APPROVED** [review]

**Review by ABingHuang — APPROVED** [review]

**Comment by Youngwb** [comment]
@mergify backport branch-3.2

**Pull request merged**